### PR TITLE
Change dragging states before triggering end event

### PR DESCRIPTION
**Pull request by frenzibyte**
Matches behaviour as hover lose and focus lose, where the property would have its state changed before triggering the event.

Have checked for any usages of this property, none of them needs any action.

# vNext

## Both `Drawable.IsDragged` and `InputManager.DraggedDrawable` now change their state before triggering event

Both `Drawable.IsDragged` and `InputManager.DraggedDrawable` now change their state before the `DragEndEvent` is triggered, rather than after, matching other similar events like `FocusLostEvent` and `HoverLostEvent`.

### CODE REVIEW

File: osu.Framework/Input/MouseButtonEventManager.cs [hunk 1]
```diff
@@ -208,9 +208,9 @@ private void handleDragEnd(InputState state)
 
             if (DraggedDrawable == null) return;
 
+            DraggedDrawable.IsDragged = false;
             PropagateButtonEvent(new[] { DraggedDrawable }, new DragEndEvent(state, Button, MouseDownPosition));
 
-            DraggedDrawable.IsDragged = false;
             DraggedDrawable = null;
```
Comment (bdach) [inline osu.Framework/Input/MouseButtonEventManager.cs:214]: This should also be unset before the event is propagated. Compare focus, which stores the previous `FocusedDrawable` to a local and only then triggers the event. Hover does similar.

This should also be included in the breaking changes note.

**Review by bdach — CHANGES_REQUESTED** [review]

File: osu.Framework/Input/MouseButtonEventManager.cs (hunk 1 above)
Comment (frenzibyte) [reply osu.Framework/Input/MouseButtonEventManager.cs:214]: Ah, missed to notice that.

File: osu.Framework/Input/MouseButtonEventManager.cs [hunk 2]
```diff
@@ -206,12 +206,15 @@ private void handleDragEnd(InputState state)
         {
             DragStarted = false;
 
-            if (DraggedDrawable == null) return;
```
Comment (bdach) [inline osu.Framework/Input/MouseButtonEventManager.cs:209]: not sure why this guard went away. something like the following reads much better and has no nesting/null checks:

```csharp
if (DraggedDrawable == null) return;

var previousDragged = DraggedDrawable;
previousDragged.IsDragged = false;
DraggedDrawable = null;

PropagateButtonEvent(new[] { DraggedDrawable }, new DragEndEvent(state, Button, MouseDownPosition));
```

but i guess this variant also works...

Comment (frenzibyte) [reply osu.Framework/Input/MouseButtonEventManager.cs:209]: Went with the same way focus lose logic does it, felt good to me at first, but I think yours sound much better.

**Review by bdach — APPROVED** [review]

**Review by smoogipoo — APPROVED** [review]
Makes sense

**Review by peppy — APPROVED** [review]

**Pull request merged**